### ISSUE TITLE
GlideinRequestManifests errors out py2.7/1.1.2 modules if DESIRED_Sites is NaN

### ISSUE THREAD

**Issue by StevenCTimm** [HEPCloud/decisionengine_modules#227]
2020-05-13 22:06:02,447 - root - TaskManager - 13414 - GlideinRequestManifests - ERROR - exception from transform GlideinRequestManifests : cannot index with vector containing NA / NaN values
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/decisionengine/framework/taskmanager/TaskManager.py", line 397, in run_transform
    data = transform.worker.transform(data_block)
  File "/usr/lib/python2.7/site-packages/decisionengine_modules/glideinwms/transforms/glidein_requests.py", line 167, in transform
    job_filter=job_query, fom_entries=fom_entries)
  File "/usr/lib/python2.7/site-packages/decisionengine_modules/glideinwms/glide_frontend_element.py", line 1289, in generate_glidein_requests
    factory_globals, job_filter=job_filter)
  File "/usr/lib/python2.7/site-packages/decisionengine_modules/glideinwms/glide_frontend_element.py", line 1346, in generate_glidein_requests_one
    jobs_df = jobs_df.query(job_filter)
  File "/usr/lib64/python2.7/site-packages/pandas/core/frame.py", line 2854, in query
    new_data = self[res]
  File "/usr/lib64/python2.7/site-packages/pandas/core/frame.py", line 2682, in __getitem__
    return self._getitem_array(key)
  File "/usr/lib64/python2.7/site-packages/pandas/core/frame.py", line 2709, in _getitem_array
    if com.is_bool_indexer(key):
  File "/usr/lib64/python2.7/site-packages/pandas/core/common.py", line 107, in is_bool_indexer
    raise ValueError('cannot index with vector containing '
ValueError: cannot index with vector containing NA / NaN values

----

Before the upgrade to 1.1.2 this used to work.  We have a stock sample job without DESIRED_Sites set and there was no error.

**Comment by StevenCTimm** [comment]
Was this fix supposed to have been in 1.1.3 branch?

**Comment by StevenCTimm** [comment]
This problem persists in 1.1.3.  I have further discovered that to trigger the error that you have to have some group in job_classification.conf that has a selection criteria of 
DESIRED_Sites.str.contains('T3_US_NERSC'). or some similar site.
at first I did not see the problem on fermicloud155 but then I added those groups and it happened.

also to see the error you have to submit jobs with no DESIRED_Sites field at all.
I will shortly check these test scripts into git.

**Comment by DmitryLitvintsev** [comment]
This issue has not been fixed or even attempted to be fixed in 1.1.3

**Comment by StevenCTimm** [comment]
diff /usr/lib/python2.7/site-packages/decisionengine_modules/glideinwms/glide_frontend_element.py /usr/lib/python2.7/site-packages/decisionengine_modules/glideinwms/glide_frontend_element.py~
1365c1365
<             jobs_df = jobs_df.fillna('').query(job_filter)
---
>             jobs_df = jobs_df.query(job_filter)

Dmitry did the one line patch on fermicloud155.. I am running with this now.
The routine hasn't error out yet.

However there is a related error in job_classification.log that we need to understand
that also happens in the case where DESIRED_Sites = nan

2020-06-10 21:19:06,095 - root - job_clustering - 30482 - JobClustering - ERROR - Unable to calculate totals from job manifests, may have missing classads or incorrect classad names

I want to get to the bottom of that as well.

**Comment by DmitryLitvintsev** [comment]
a stack trace could be helpful

**Comment by StevenCTimm** [comment]
The stack trace is above at the beginning of the ticket.

The other error that I just put above, "Unable to calculate totals from job manifests" is part of an 
Except clause in glideinwms/transforms/job_clustering.py 
IT could be one of three different catches of the exception at lines 105, 109, 113 of that routine.
Line 94  of that same routine is 

  df_q = df_full_q.query(self.job_q_expr)

which since it is running on the very same data frame as the one above, should not be surprising that it also throws an exception...likely to be a ValueError as the one above.

**Comment by StevenCTimm** [comment]
As mentioned above, fermicloud155 with the one-line patch as above is not throwing an exception now, but neither is it successfully completing the resource request channel.. I am still investigating
and want to get it sorted out.

The one-line patch seems to have fixed the "Unable to calculate totals" thing too, and I don't understand that, it should not have done so.

I want to get a full chain through on fermicloud155 before I say ok to deploy the patch as is.

**Comment by DmitryLitvintsev** [comment]
I thought that

```
2020-06-10 21:19:06,095 - root - job_clustering - 30482 - JobClustering - ERROR - Unable to calculate totals from job manifests, may have missing classads or incorrect classad names
```
is  new problem. The stack trace at the very top addressed by the one line change

**Comment by StevenCTimm** [comment]
No--both problems were happening all the time simultaneously.. still trying to get this to the bottom of this.

**Comment by StevenCTimm** [comment]
OK so a bit of clarity:
The error in glide_frontend_element.py happens if there are any DESIRED_Sites NaN in the job_manifest datablock 

The error in job_clustering.py only happens if there are actually groups in the job_classification.conf that have a job query expression that uses DESIRED_Sites and any resources that actually match the resource selection expression.

I think the right thing to do is to add the fillna option to the line in job_clustering.py similar as to what was done in the glide_frontend_element.py.  But let me do it.. there are still three steps I have to test before that.
I have spent most of the time up to now to get fermicloud155 in a state where it can successfully request glideins from the factory--up until now it could not do so.

Once it request glideins in the current state, then I will re-enable the groups that have DESIRED_Sites, see if I get an 
"unable to calculate totals" error.. 
if so, then patch that routine, 
see if that fixes it.

then apply the same to hepcsvc03 and see if it fixes it there too.

It has taken a while to get causality but I think we are almost there.

**Comment by StevenCTimm** [comment]
Now have successfully gotten glideins from fermicloud155 to the factory with the patch in.
Now adding the other groups back in job_classification.conf.

**Comment by StevenCTimm** [comment]
now introduced the 6 groups with DESIRED_Sites.str.contains() back into the job_classification.conf on fermicloud155 "Unable to calculate totals" error came right back
(and a job_clusters block of all zeros too).

Patched line 94 of job_clustering.py as discussed above, restarted the DE.

"Unable to calculate totals" error is now gone on fermicloud155, resources successfully being requested from cmssrv280 factory.

`diff job_clustering.py /tmp/job_clustering.py 
87c87
< #            df_q = df_full_q.query(self.job_q_expr)
---
> #            df_q = df_full_q.fillna('').query(self.job_q_expr)
95c95
<             df_q = df_full_q.query(self.job_q_expr)
---
>             df_q = df_full_q.fillna('').query(self.job_q_expr)`

**Comment by StevenCTimm** [comment]
I don't know why the diff looks so crazy above.
anyway you get the idea

both fermicloud155 and hepcsvc03 are stably processing a large queue of jobs that do not have DESIRED_Sites set.. if this holds stable for a few hours I will say it is time to make the patch.
and will even review it speedily if I can.

**Comment by DmitryLitvintsev** [comment]
https://fermicloud140.fnal.gov/reviews/r/154/

**Comment by StevenCTimm** [comment]
It is holding steady.
The same patch also needs to be applied to the 1.2 branch.  We will have the same trouble there.
and in the case of the job_clustering.py file we can already see it.

**Comment by StevenCTimm** [comment]
This is now fixed in the 1.1.4 rpm that has just been released.  This ticket can be closed.
